elfo-core/src/restarting/backoff.rs

```diff
@@ -0,0 +1,135 @@
+use std::time::Duration;
+
+use crate::RestartParams;
+use elfo_utils::time::Instant;
+
+pub(crate) struct RestartBackoff {
+    start_time: Instant,
+    restart_count: u64,
+    power: u64,
+}
+
+impl Default for RestartBackoff {
+    fn default() -> Self {
+        Self {
+            start_time: Instant::now(),
+            restart_count: 0,
+            power: 0,
+        }
+    }
+}
+
+impl RestartBackoff {
+    pub(crate) fn start(&mut self) {
+        self.start_time = Instant::now();
+    }
+
+    pub(crate) fn next(&mut self, params: &RestartParams) -> Option<Duration> {
+        // If an actor is alive enough time, reset the backoff.
+        if self.start_time.elapsed() >= params.auto_reset {
+            self.restart_count = 1;
+            self.power = 0;
+            return Some(Duration::ZERO);
+        }
+        self.restart_count += 1;
+
+        if self.restart_count > params.max_retries.get() {
+            return None;
+        }
+
+        let delay = (params.min_backoff.as_secs_f64() * params.factor.powf(self.power as f64))
+            .clamp(
+                params.min_backoff.as_secs_f64(),
+                params.max_backoff.as_secs_f64(),
+            );
+
+        self.power += 1;
+        // Check for overflow, if overflow is detected set the current delay to maximal
+        let delay = if delay.is_finite() {
+            Duration::from_secs_f64(delay)
```

Comment:
Add a comment that factor is checked against negativity and therefore it doesn't lead to panic